File: ledger/ledger-consensual/src/main/java/net/corda/v5/ledger/consensual/ConsensualLedgerService.java
package net.corda.v5.ledger.consensual;

import net.corda.v5.application.messaging.FlowSession;
import net.corda.v5.base.annotations.DoNotImplement;
import net.corda.v5.base.annotations.Suspendable;
import net.corda.v5.crypto.SecureHash;
import net.corda.v5.ledger.consensual.transaction.ConsensualLedgerTransaction;
import net.corda.v5.ledger.consensual.transaction.ConsensualSignedTransaction;
import net.corda.v5.ledger.consensual.transaction.ConsensualTransactionBuilder;
import net.corda.v5.ledger.consensual.transaction.ConsensualTransactionValidator;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

/**
 * Defines the consensual ledger service.
 */
@DoNotImplement
public interface ConsensualLedgerService {

    /**
     * Gets a {@link ConsensualTransactionBuilder} instance.
     *
     * @return Returns a new {@link ConsensualTransactionBuilder} instance.
     */
    @NotNull
    @Suspendable
    ConsensualTransactionBuilder getTransactionBuilder();

    /**
     * Finds a {@link ConsensualSignedTransaction} in the vault by its transaction ID.
     *
     * @param id The transaction ID of the {@link ConsensualSignedTransaction} to find in the vault.
     * @return Returns the {@link ConsensualSignedTransaction} if it has been recorded, or null if the transaction could not be found.
     */
    @Nullable
    @Suspendable
    ConsensualSignedTransaction findSignedTransaction(@NotNull SecureHash id);

    /**
     * Finds a {@link ConsensualLedgerTransaction} in the vault by its transaction ID.
     *
     * @param id The transaction ID of the {@link ConsensualLedgerTransaction} to find in the vault.
     * @return Returns the {@link ConsensualLedgerTransaction} if it has been recorded, or null if the transaction could not be found.
     */
    @Nullable
    @Suspendable
    ConsensualLedgerTransaction findLedgerTransaction(@NotNull SecureHash id);

    /**
     * Finalizes a transaction by collecting any remaining required signatures from counter-parties, and broadcasts the
     * fully signed transaction to all participants involved in the transaction to be recorded in the vault.
     *
     * @param transaction The transaction to finalize.
     * @param sessions The sessions representing the counter-party participants of the transaction.
     * @return Returns the fully signed and recorded transaction.
     */
    @NotNull
    @Suspendable
    ConsensualSignedTransaction finalize(
            @NotNull ConsensualSignedTransaction transaction,
            @NotNull List<FlowSession> sessions
    );

    /**
     * Verifies, signs and records the fully signed {@link ConsensualSignedTransaction}.
     *
     * @param session The session from which the {@link ConsensualSignedTransaction} was received.
     * @param validator Validates the received {@link ConsensualSignedTransaction}.
     * @return Returns the fully signed {@link ConsensualSignedTransaction}.
     */
    @NotNull
    @Suspendable
    ConsensualSignedTransaction receiveFinality(
            @NotNull FlowSession session,
            @NotNull ConsensualTransactionValidator validator
    );
}

File: ledger/ledger-consensual/src/main/kotlin/net/corda/v5/ledger/consensual/ConsensualLedgerService.kt
package net.corda.v5.ledger.consensual

import net.corda.v5.application.messaging.FlowSession
import net.corda.v5.base.annotations.DoNotImplement
import net.corda.v5.base.annotations.Suspendable
import net.corda.v5.crypto.SecureHash
import net.corda.v5.ledger.consensual.transaction.ConsensualLedgerTransaction
import net.corda.v5.ledger.consensual.transaction.ConsensualSignedTransaction
import net.corda.v5.ledger.consensual.transaction.ConsensualTransactionValidator
import net.corda.v5.ledger.consensual.transaction.ConsensualTransactionBuilder

/**
 * It provides access to the different Consensual Ledger Services
 */
@DoNotImplement
interface ConsensualLedgerService {
    /**
     * Gets a Consensual transaction builder.
     *
     * @return Returns a new [ConsensualTransactionBuilder] instance.
     */
    @Suspendable
    fun getTransactionBuilder(): ConsensualTransactionBuilder

    /**
     * Finds a transaction by id in the vault and returns it
     *
     * @param id The id of the transaction to find
     *
     * @return The signed transaction, if it has been recorded previously. Null if not found.
     */
    @Suspendable
    fun findSignedTransaction(id: SecureHash): ConsensualSignedTransaction?

    /**
     * Finds a transaction by id in the vault, resolves it to a ledger transaction and returns it
     *
     * @param id The id of the transaction to find
     *
     * @return The ledger transaction, if it has been recorded previously. Null if not found.
     */
    @Suspendable
    fun findLedgerTransaction(id: SecureHash): ConsensualLedgerTransaction?

    /**
     * Collects signatures, records and broadcasts to involved peers a [ConsensualSignedTransaction].
     *
     * @param signedTransaction The [ConsensualSignedTransaction] to finalise and recorded locally and with peer [sessions].
     * @param sessions The [FlowSession]s of the peers involved in the transaction.
     *
     * @return The fully signed [ConsensualSignedTransaction] that was recorded.
     */
    @Suspendable
    fun finalize(
        signedTransaction: ConsensualSignedTransaction,
        sessions: List<FlowSession>
    ): ConsensualSignedTransaction

    /**
     * Verifies, signs and records a [ConsensualSignedTransaction].
     *
     * @param session The [FlowSession] to receive the [ConsensualSignedTransaction] from.
     * @param validator Validates the received [ConsensualSignedTransaction].
     *
     * @return The fully signed [ConsensualSignedTransaction] that was received and recorded.
     */
    @Suspendable
    fun receiveFinality(
        session: FlowSession,
        validator: ConsensualTransactionValidator
    ): ConsensualSignedTransaction
}